### ISSUE TITLE
[HOLD #147480][$40,000] React Native ScrollView bug: maintainVisibleContentPosition

### ISSUE THREAD

**Issue by roryabraham** [Expensify/App#7925]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
Here is a [very minimal reproduction](https://github.com/facebook/react-native/pull/33184) of this bug in the [rn-tester sample project](https://github.com/facebook/react-native/tree/main/packages/rn-tester), that's not dependent upon any new code in the React Native codebase. Run the rn-tester app, find the `ScrollView` example, and observe the following behavior:

If `minIndexForVisible` is `0`, then the scroll position will be maintained iff:

1. The first item in the list is not visible, _AND_
2. The current `contentOffset` is at least `y`, where `y` is the height of the new content being prepended

Similarly, if `minIndexForVisible` is `1`, then the scroll position will be maintained iff:

1. The _second_ item in the list is not visible, _AND_
2. The current `contentOffset` is at least `y`, where `y` is the height of the new content being prepended

And so on... If either of those conditions are not met, the scroll position will not be maintained.

## Expected Result:
As long as the list has enough content that it is scrollable, the `contentOffset` should be adjusted such that the scroll position is maintained when new items are added to the start or end of the list.

## Actual Result:
The `contentOffset` is not adjusted, and the scroll position is not maintained.

### Background
How does the `maintainVisibleContentPosition` prop work?

At a high level, when [new UI elements are added to the start of the list](https://github.com/facebook/react-native/blob/8bd3edec88148d0ab1f225d2119435681fbbba33/React/Views/ScrollView/RCTScrollView.m#L902), React Native will:

1. [Measure the position of the first visible item](https://github.com/facebook/react-native/blob/8bd3edec88148d0ab1f225d2119435681fbbba33/React/Views/ScrollView/RCTScrollView.m#L926-L927) before the new items are added
2. Add the new items to the view.
3. [Measure the difference in position](https://github.com/facebook/react-native/blob/8bd3edec88148d0ab1f225d2119435681fbbba33/React/Views/ScrollView/RCTScrollView.m#L953-L954) between the first visible item found in step 1 with its new position.
4. [Increase the `contentOffset` of the scroll container](https://github.com/facebook/react-native/blob/8bd3edec88148d0ab1f225d2119435681fbbba33/React/Views/ScrollView/RCTScrollView.m#L964) by the amount calculated in the previous step, such that:
    1. The same first item is visible before and after adding the new items to the list, and
    2. All the newly-prepended items are out of view, and further towards the start of the list.

However, this prop does not work consistently – sometimes in step 3 the [difference in position](https://github.com/facebook/react-native/blob/8bd3edec88148d0ab1f225d2119435681fbbba33/React/Views/ScrollView/RCTScrollView.m#L954-L955) is incorrectly calculated to be zero. Furthermore, we have noticed that this seems only to happen consistently when the content length of newly-prepended list items is long.

### Motivation
For a few months now, we have been endeavoring to get a working solution for a bidirectional-scrolling virtualized list in React Native. After working through many potential solutions, we have come very close to a working solution directly in React Native's VirtualizedList through the code in [this PR](https://github.com/Expensify/react-native/pull/7). However, after lots of debugging we determined that the issues we were seeing weren't caused by the JS / `VirtualizedList` at all, but instead by this bug in [ScrollView's `maintainVisibleContentPosition` prop](https://github.com/facebook/react-native/blob/1814bffcc2139b9d7ff4660e5ec184639c08f6ec/Libraries/Components/ScrollView/ScrollView.js#L265-L292), which is implemented in the native layer.

The _result_ of this bug is that our implementation of the `onStartReached` prop in `VirtualizedList` suffers from the following issue:

1. When you reach the start of the list (`contentOffset` of `0`), `onStartReached` is called.
2. The callback to `onStartReached` prepends new items into the list.
3. `maintainVisibleContentPosition` fails to update the `contentOffset` to account for those new list items.
4. The new list items are rendered, but the `contentOffset` is still `0`, so the list position jumps to the start of the new content.
5. Because the `contentOffset` is `0`, `onStartReached` is called again, and we get an infinite loop (at least, until there's no more content to load).

### Android considerations
Another important piece of information is that the `maintainVisibleContentPosition` prop is not yet available on Android (implementation [in progress](https://github.com/facebook/react-native/pull/29466)). We have examined the in-progress Android implementation and found that it is very similar to the iOS one, and likely shares the same problem.

For the sake of this issue, the scope is focused on `iOS`, but we believe that the solution in one platform will be applicable in the other.

### Potential cause
According to [review comments from a Meta engineer](https://github.com/facebook/react-native/pull/29466/files#r779111410), this bug is likely caused by a race condition between the items being added to the list and content offset being adjusted.

They also suggest implementing a binary search for the first visible item, which seems like it might improve the issue, but (in my opinion) is unlikely to resolve the race condition entirely.

### Evidence of potential race condition?
In the FlatList example linked above, if you tweak [these parameters](https://github.com/Expensify/react-native/pull/7/files#diff-e30adf9ac66888db4fe1a9c256590a37677724935458dae6b7a61b99331e292eR46-R48) as follows:

```js
const PAGE_SIZE = 10;
const INITIAL_PAGE_OFFSET = 50;
const NUM_PAGES = 100;
```

The problem is mitigated but not completely solved (a few pages load before `maintainVisibleContentPosition` seems to "catch up" and function as expected). This hints that the problem may indeed be a race condition as suggested above.

### `autoScrollToTopThreshold` wonkiness

According to the React Native documentation:

> The optional autoscrollToTopThreshold can be used to make the content automatically scroll to the top after making the adjustment if the user was within the threshold of the top before the adjustment was made.

This suggests that with an `autoScrollToTopThreshold` of `0`, then no auto-scrolling should occur if you have a non-zero `contentOffset` before new items are appended to the list. We have observed that this is not the case by:

1. Scrolling down a few pixels (without taking the `minIndexForVisible` item out of view)
2. Prepending an item.

Despite having an `autoScrollToTopThreshold` of `0` and a non-zero `contentOffset`, the `ScrollView` auto-scrolls to the top.

Interestingly, [this particular `ScrollView` example](https://github.com/facebook/react-native/pull/33184) listed in the reproduction steps can be fixed by removing the `autoScrollToTopThreshold` parameter entirely. While this might be a hint at how to solve this, _it does not seem to be a viable solution for us_. Even without an `autoScrollToTopThreshold` parameter, the same problem occurs in [this FlatList example](https://github.com/Expensify/react-native/pull/7). It's unclear why removing the `autoScrollToTopThreshold` parameter fixes the problem, but setting a value of `0` does not. 🤔

## Workaround:
While there may be workarounds possible via hacks in the JS code involving `setTimeout` or extra calls to `scrollToIndex`/`scrollToOffset`, these would not solve the root problem. In order to have a proposal accepted, it must fix the problem in the React Native codebase itself, probably in the native layer.

## Platform:
Right now, this problem has been _confirmed_ on iOS. It very likely exists on Android as well in [this implementation](https://github.com/facebook/react-native/pull/29466/). We are working on confirming the issue there, and will be following the progress of that pull request.

For the scope of this issue, we'll only require a fix in iOS _or_ Android (preferably iOS), submitted as a PR against the our react-native fork: https://github.com/Expensify/react-native. Applying the same fix in the other platform should be comparatively easy and can be treated as a follow-up.

---

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01fe321bebf9b78f69</li>
        <li>Upwork Job ID: 1606337510652706816</li>
        <li>Last Price Increase: 2022-12-23</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @kadiealexander (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by roryabraham** [comment]
@kadiealexander reassigning to @mallenexpensify because we discussed this 1:1

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @parasharrajat (`Exported`)

**Comment by MelvinBot** [comment]
Current assignee @roryabraham is eligible for the Exported assigner, not assigning anyone new.

**Comment by mallenexpensify** [comment]
We live! https://www.upwork.com/jobs/~01620b5ab4dee0a9f0

**Comment by zoontek** [comment]
I'm not sure I get the issue correctly since, when setting:

```jsx
<ScrollView
  automaticallyAdjustContentInsets={false}
  maintainVisibleContentPosition={{ minIndexForVisible: 0 }}
  // …
```

`contentOffset` is indeed adjusted, scroll position maintained:

https://user-images.githubusercontent.com/1902323/155970666-d1294c7f-c624-432a-9530-87a704d81bc3.MP4

---

```jsx
<ScrollView
  automaticallyAdjustContentInsets={false}
  maintainVisibleContentPosition={{
    minIndexForVisible: 0,
    autoscrollToTopThreshold: 0,
  }}
```

`contentOffset` is adjusted, scroll position is updated if `minIndexForVisible` element +- `autoscrollToTopThreshold` is visible (approximately, sure).

https://user-images.githubusercontent.com/1902323/155971211-39613016-2901-498e-858e-2df52f184b9e.MP4

---

But it really is broken with:

```jsx
<ScrollView
  automaticallyAdjustContentInsets={false}
  maintainVisibleContentPosition={{
    minIndexForVisible: 1,
    autoscrollToTopThreshold: 1,
  }}
```

With `contentOffset` being incorrect when removing elements. But it's not the purpose of this issue.

https://user-images.githubusercontent.com/1902323/155971475-b4e1b7ec-0bbc-4ad1-9875-7c333d6fa043.MP4

@roryabraham Maybe I'm missing something?

**Comment by roryabraham** [comment]
@zoontek Thanks for following up. You may have missed this piece of the issue description:

> Interestingly, https://github.com/facebook/react-native/pull/33184 listed in the reproduction steps can be fixed by removing the autoScrollToTopThreshold parameter entirely. While this might be a hint at how to solve this, it does not seem to be a viable solution for us. Even without an autoScrollToTopThreshold parameter, the same problem occurs in https://github.com/Expensify/react-native/pull/7. It's unclear why removing the autoScrollToTopThreshold parameter fixes the problem, but setting a value of 0 does not. 🤔

**Comment by zoontek** [comment]
@roryabraham No, I read it, but as I understand setting any `autoScrollToTopThreshold` value will activate auto scroll to top if the given index is visible. Which it does (?)

**Comment by roryabraham** [comment]
Hi @zoontek – so you're saying that `autoScrollToTopThreshold` doesn't represent a pixel offset, but an offset represented in visual items? Interesting, as I did not glean that from the RN docs or implementation.

Nonetheless, I think this is still a valid bug report in that `maintainVisibleContentPosition` does not work consistently / appears subject to some race condition(s). Perhaps our minimal reproduction example is faulty / needs improvement, but the bug we observed is clearly reproducible in [this PR branch](https://github.com/Expensify/react-native/pull/7), and I believe that bug surfaces in the native layer at the location described in the issue template.

**Comment by roryabraham** [comment]
@mallenexpensify Let's double this.

**Comment by mallenexpensify** [comment]
Price doubled to $20,000

**Comment by azimgd** [comment]
@roryabraham could you take a look at this code please: https://github.com/azimgd/MVCPExample/

Here is the demo which maintains visible content position in both directions, is that the actual result you are looking for in this issue?:
https://user-images.githubusercontent.com/4882133/157487057-ad6e4acf-c13b-4efe-bf07-68a47793c309.mov

[CODEFIX DIFF](https://github.com/azimgd/MVCPExample/commit/6f02aca48753d236e62f48972583fae0e3f438b1#diff-53de7701df96e2c957bbf9dd73edd44ba365f90b4728ec5ec2a08baf86520955R263-R292)

```diff
       CGRect newFrame = self->_firstVisibleView.frame;
-      CGFloat deltaY = newFrame.origin.y - self->_prevFirstVisibleFrame.origin.y;
-      if (ABS(deltaY) > 0.1) {
+      CGFloat deltaY = self->_scrollView.contentSize.height - self->_prevContentSizeHeight;
+        
+      if (ABS(deltaY) > 0.1 && newFrame.origin.y != self->_prevFirstVisibleFrame.origin.y) {
```

Changing `minIndexForVisible` is not exactly focusing on that index right now, but I will get that working once I validate the exact requirements.

**Comment by Sarveshwins** [comment]
could you take a look at this code please: https://github.com/Sarveshwins/ScrollViewbug

Here is the demo which maintains visible content position in both directions.

https://user-images.githubusercontent.com/82214917/157656995-7e47c968-2656-479a-a1d0-22fc2b06491b.mov

  const [sizeOfHeight, setsizeOfHeight] = useState(40);
  const [bottomSizeOfHeight, setBottomSizeOfHeight] = useState(40);
  const ChangeHeight = () => {
    setsizeOfHeight(prevState => {
      return prevState === 40 ? 90 : 40;
    });
  };
  const ChangeBottomHeight = () => {
    setBottomSizeOfHeight(prevState => {
      return prevState === 40 ? 90 : 40;
    });
  };
  function getRandomColor() {
    let letters = '0123456789ABCDEF';
    let color = '#';
    for (let i = 0; i < 6; i++) {
      color += letters[Math.floor(Math.random() * 16)];
    }
    return color;
  }
  const generate = (size = 10) => {
    const rows = new Array(size).fill(true).map((item, index) => ({
      key: index,
      size: Math.round(Math.random() * 100),
      color: getRandomColor(),
    }));
    return rows;
  };
  const [dataForVisible, setdataForVisible] = useState(generate());
  const OnAppend = () => {
    setdataForVisible(state => {
      return [
        {
          size: Math.round(Math.random() * 100),
          color: getRandomColor(),
          widthsize: sizeOfHeight,
        },

        ...state,
      ];
    });
  };
  const onFilter = () => {
    setdataForVisible(state => {
      return state.filter((item, index) => index != 0);
    });
  };
  const addAtBottom = () => {
    setdataForVisible(state => {
      return [
        ...state,
        {
          size: Math.round(Math.random() * 100),
          color: getRandomColor(),
          widthsize: bottomSizeOfHeight,
        },
      ];
    });
  };

  const DeletAtBottom = () => {
    setdataForVisible(state => {
      return state.filter((item, index) => index != state.length - 1);
    });
  };

**Comment by parasharrajat** [comment]
Looking at proposals shortly...

**Comment by zoontek** [comment]
@roryabraham I took a bit of time on this one, made a deep dive in the code to understand it better (as I find the documentation not that clear)

```jsx
maintainVisibleContentPosition={{
  minIndexForVisible: 0,
}}
```

Scroll is correctly preserved since the first entirely visible view is indeed at index >= 0:

https://user-images.githubusercontent.com/1902323/157931513-0dbb6dcd-fda3-49cd-96c1-c589c062a2e4.mp4

Now with `autoscrollToTopThreshold`. As we can see, each item here has a 37dp height:

<img src="https://user-images.githubusercontent.com/1902323/157931742-5962601e-3d8f-4610-ae99-f6f07c1cfaa5.png" width="300px" />

Which means it should scroll to top when threshold (scrollview contentOffset.y) is < 37dp:

```jsx
maintainVisibleContentPosition={{
  minIndexForVisible: 0,
  autoscrollToTopThreshold: 37,
}}
```

https://user-images.githubusercontent.com/1902323/157931957-eabf77a2-6fd7-412b-be3d-cae5dad7b9a7.mp4

But keep in mind that adding / removing items will update indexes (`minIndexForVisible` item will be updated) and `autoscrollToTopThreshold` limit:

<img width="1082" alt="autoscrollToTopThreshold schema" src="https://user-images.githubusercontent.com/1902323/158015533-632ec83c-1e03-4d8e-94d3-4ddf1fdc2517.png">

https://user-images.githubusercontent.com/1902323/158015565-d91fd7f1-5c88-425f-ba18-570e8e3c5c5f.mp4

---

Now, several remarks:

1. Should scroll position be preserved if the item at `minIndexForVisible` is not visible? Currently it is, setting it to `10` for example will maintain scroll position even if at initial list state, this item is not visible yet. (we should set `self->_prevFirstVisibleFrame` and `self->_firstVisibleView` to `nil` to avoid maintaining position in such cases - but I understand why it has been ignored - `minIndexForVisible` should be se set to `0` or `1` at max, allowing you to handle a loader which will always be the first item in your list, even if you prepending a lot of new items).

2. `ABS(deltaY)` alone is a mistake. `deltaY > 0.1` (adding item) and `deltaY < -0.1` (removing item) should be handled in a different way since the adjustment might not be the same.

Quick example of the current behaviour:

```jsx
maintainVisibleContentPosition={{
  minIndexForVisible: 1,
}}
```

https://user-images.githubusercontent.com/1902323/158016088-bd654b22-e732-42ed-bcf8-e4831700166d.mp4

Improved behavior:

https://user-images.githubusercontent.com/1902323/158016289-2534597b-f1bf-4b33-9c5c-bbb2c9cacc22.mp4

The current code (I removed horizontal scroll handling for example clarity, but the logic is the same):

```objc
- (void)uiManagerWillPerformMounting:(RCTUIManager *)manager
{
  RCTAssertUIManagerQueue();

  [manager prependUIBlock:^(__unused RCTUIManager *uiManager,
                            __unused NSDictionary<NSNumber *, UIView *> *viewRegistry) {
    NSUInteger minIndexForVisible = [self->_maintainVisibleContentPosition[@"minIndexForVisible"] integerValue];
    NSUInteger itemCount = self->_contentView.subviews.count;
    CGFloat scrollViewContentOffsetY = self->_scrollView.contentOffset.y;

    for (NSUInteger index = minIndexForVisible; index < itemCount; ++index) {
      // Find the first entirely visible view. This must be done after we update the content offset
      // or it will tend to grab rows that were made visible by the shift in position
      UIView *subview = self->_contentView.subviews[index];
      CGFloat subviewOriginY = subview.frame.origin.y;

      if (subviewOriginY >= scrollViewContentOffsetY || index == itemCount - 1) {
        self->_prevFirstVisibleFrame = subview.frame;
        self->_firstVisibleView = subview;
        break;
      }
    }
  }];

  [manager addUIBlock:^(__unused RCTUIManager *uiManager,
                        __unused NSDictionary<NSNumber *, UIView *> *viewRegistry) {
    if (self->_maintainVisibleContentPosition == nil) {
      return; // The prop might have changed in the previous UIBlocks, so need to abort here.
    }

    NSNumber *autoscrollToTopThreshold = self->_maintainVisibleContentPosition[@"autoscrollToTopThreshold"];
    CGRect newFrame = self->_firstVisibleView.frame;
    CGFloat deltaY = newFrame.origin.y - self->_prevFirstVisibleFrame.origin.y;

    if (ABS(deltaY) > 0.1) {
      // Handle items removal properly
      self->_scrollView.contentOffset = deltaY < -0.1 && self->_scrollView.contentOffset.y < newFrame.size.height
        ? CGPointMake(self->_scrollView.contentOffset.x, 0)
        : CGPointMake(self->_scrollView.contentOffset.x, self->_scrollView.contentOffset.y + deltaY);

      if (autoscrollToTopThreshold != nil) {
        // If the offset WAS within the threshold of the start, animate to the start.
        if (self->_scrollView.contentOffset.y - deltaY <= [autoscrollToTopThreshold integerValue]) {
          [self scrollToOffset:CGPointMake(self->_scrollView.contentOffset.x, 0) animated:YES];
        }
      }
    }
  }];
}
````

And the repository for the JS part: https://github.com/zoontek/ScrollViewIssue (you will have to manually replace `uiManagerWillPerformMounting` in `Pods > Development Pods > React-Core > Default > Views > ScrollView > RCTScrollView.m`).

Again: there is no real changes here (except to improve items removal handling). This API has a few quirks that can be fixed, but given my explanations (mainly about `autoscrollToTopThreshold` behaviour), does the initial issue described in the ticket really exists?
Maybe would you want to fix `minIndexForVisible` and `autoscrollToTopThreshold` in time? By that, I mean that once the content of the `ScrollView` is "stable" **and** the `maintainVisibleContentPosition` prop is set, we keep a delta of the prepended items index and their size to adjust `minIndexForVisible` and `autoscrollToTopThreshold`? That would be a bit weird, but it's doable.


<img width="1273" alt="with deltas schema" src="https://user-images.githubusercontent.com/1902323/158016959-0b56bb36-b0e6-46f2-a7fb-712b1dfb3f09.png">

**Comment by parasharrajat** [comment]
@Sarveshwins Sorry but requirements mention that this issue needs to be fixed on react-native source.

**Comment by parasharrajat** [comment]
@zoontek I can't access mentioned repo.
![image](https://user-images.githubusercontent.com/24370807/158262399-e7bd57ba-5ef9-4db4-a15b-e5b4ad7b1d1b.png)

**Comment by zoontek** [comment]
@parasharrajat Sorry, I changed visibility to public.

**Comment by parasharrajat** [comment]
First, sorry for the delay. I was trying to set up the project on limited connectivity and trying to get grip on the problem. 

Well! The Rn docs are really bad. I have a few questions. What should be the correct behavior of `autoscrollToTopThreshold`. When this is 0 and we prepend a new item the list should scroll and show that item. right? Following that when `autoscrollToTopThreshold is 37`, prepending new items should be scrolled to view if `contentOffset.y` is less than 37. 


@azimgd I think that is the required behavior but I will confirm. In your example, I saw that you are not using `autoscrollToTopThreshold`. Issue Details explain the problem with it. Let's see if an example fixes it. 
![image](https://user-images.githubusercontent.com/24370807/158274960-da9d37bb-8077-433c-a0b2-7a3dd46d6d83.png)


> Again: there are no real changes here (except to improve items removal handling). The original API has a few quirks that can be fixed, but given my explanations (mainly about autoscrollToTopThreshold behavior), does the initial issue described in the ticket really exists?

 @zoontek I don't have a proper answer to it. But I think that the issue in the above-attached image should not happen. 

Also, this problem was initially faced on https://github.com/Expensify/react-native/pull/7. Maybe changes on that PR can help you answer a few questions.

**Comment by azimgd** [comment]
> Interestingly, https://github.com/facebook/react-native/pull/33184 listed in the reproduction steps can be fixed by removing the autoScrollToTopThreshold parameter entirely. While this might be a hint at how to solve this, it does not seem to be a viable solution for us. Even without an autoScrollToTopThreshold parameter, the same problem occurs in https://github.com/Expensify/react-native/pull/7. It's unclear why removing the autoScrollToTopThreshold parameter fixes the problem, but setting a value of 0 does not. 🤔

The answer is `<=` comparison operator:
```js
// RCTScrollView.m

/**
 * This is why removing the autoScrollToTopThreshold parameter entirely works
 */
if (autoscrollThreshold != nil) {
  /**
   * having autoscrollThreshold === 0; will evaluate to 0 <= 0
   */
  if (self->_scrollView.contentOffset.y - deltaY <= [autoscrollThreshold integerValue]) {
    [self scrollToOffset:CGPointMake(self->_scrollView.contentOffset.x, 0) animated:YES];
  }
}
```

Where fix for this would be:
```js
if (autoscrollThreshold != nil && autoscrollThreshold != 0) {
```

**Comment by parasharrajat** [comment]
@azimgd I was trying out your example and It was very confusing due to all the color changes and also, indexes on the item change. Could you please simplify it so that assigned indexes don't change and we can clearly see the newly added items?

It seems that list is scrolling when `minIndexforVisible: 0` on each prepend.

**Comment by azimgd** [comment]
@parasharrajat 
> I was trying out your example and It was very confusing due to all the color changes and also

Just want to make sure you run `npx patch-package`, which applies the patch into `node_modules/react-native/React/Views/ScrollView/RCTScrollView.m`


> Could you please simplify it so that assigned indexes don't change and we can clearly see the newly added items?

- Indexes are not changed, each new block has `10` items on **append** and **prepend** starting from `1..10`
- Actual requirement is that we don't want to show new items whenever `self->_contentView.subviews.count ` changed, and the scroll position should be maintained
- You can clearly see newly added items by manually scrolling in either direction


> It seems that list is scrolling when minIndexforVisible: 0 on each prepend.

I have not yet applied the fix mentioned at https://github.com/Expensify/App/issues/7925#issuecomment-1068305274

Let me know if you have any more questions and please keep in mind that we are still in the proposal stage

**Comment by parasharrajat** [comment]
> Just want to make sure you run `npx patch-package`

Aha, I missed that. Ops. 

> Let me know if you have any more questions and please keep in mind that we are still in the proposal stage

Yup. I agree. But as you asked me to test your repo and I am trying to do that.

**Comment by roryabraham** [comment]
> The answer is <= comparison operator:

Thanks @azimgd, that at least clears up some confusion I had about `autoScrollToTopThreshold`.

**Comment by azimgd** [comment]
> More precisely, your solution does not attempt to measure the location of specific subview and use that to determine the delta, and instead measures the full content size before and after new items are added when evaluating the delta.

this is correct, I was planning to implement the solution on top of that logic.

I will work on that PR, and update you soon.

**Comment by roryabraham** [comment]
Thanks @zoontek for the detailed analysis:

> Should scroll position be preserved if the item at minIndexForVisible is not visible?

I would think so, yes. Let's assume we had this `maintainVisibleContentPosition` configuration:

```js
maintainVisibleContenPosition = {
    minIndexForVisible: 0,
    autoScrollToTopThreshold: 25,
};
```

I imagine that if you are in the middle of a list (index 0 is not visible, and we are more than 25px away from the start), then new content is prepended to the list, we would maintain scroll position.

> Again: there is no real changes here (except to improve items removal handling)

Agree. Looks like those are good improvements but they don't address the issue we're facing.

> but given my explanations (mainly about autoscrollToTopThreshold behaviour), does the initial issue described in the ticket really exists?

So there was definitely a flaw in the original post for the issue. The minimal reproduction we provided (using `ScrollView` alone) does not actually reflect the issue, but was a red herring that @azimgd explained the cause for [in this comment](https://github.com/Expensify/App/issues/7925#issuecomment-1068305274).

We can remove `autoscrollToTopThreshold` entirely and the issue still exists. In order to reproduce it, try building the `rn-tester` app off of [this branch](https://github.com/Expensify/react-native/pull/7) and look at the FlatList example. By adjusting the page size you can see that the issue is only consistently reproducible when the new content has enough height, and is likely caused by some race condition.

**Comment by azimgd** [comment]
After further investigation my initial thought are following:
- `maintainVisibleContentPosition` won't behave correctly when used under FlatList because `self->_contentView.subviews.count` here https://github.com/facebook/react-native/blob/main/React/Views/ScrollView/RCTScrollView.m#L910 will have an amount of items rendered by virtualizer, not the actual array length of `data` prop passed to the `FlatList`
- Therefore the first entirely visible view that we are trying to calculate here https://github.com/facebook/react-native/blob/main/React/Views/ScrollView/RCTScrollView.m#L926 is always wrong when prepending items.
- It incorrectly calculates the position of the first visible item, resulting position to be 0px. Which becomes 0 - 0px here https://github.com/facebook/react-native/blob/main/React/Views/ScrollView/RCTScrollView.m#L954
- I'm most positive that's because items in view are shifted, due to Flatlist is getting items by index https://github.com/facebook/react-native/blob/main/Libraries/Lists/VirtualizedList.js#L1814

I will give more detailed analysis with the proposal soon.

**Comment by parasharrajat** [comment]
Thanks, @azimgd, I am eager to see the detailed analysis. You can create an external and link it here if the post is long. You can use the same to prepare a complete solution for each step you would take which can be reviewed as a whole at some point.

**Comment by parasharrajat** [comment]
Hey @fabriziobertoglio1987, 

Hope are you doing great. Sorry to ping you without notice. But I have seen your work on RN issues and really fond of your efforts to make RN better. I thought you may be interested in this job. 

Thanks.

**Comment by zoontek** [comment]
> So there was definitely a flaw in the original post for the issue. The minimal reproduction we provided (using ScrollView alone) does not actually reflect the issue

@roryabraham Maybe it doesn't really happen in your reproduction case because the issue is, indeed, tied to virtualization (could only happen inside a `FlatList` / `SectionList`?) (like @azimgd said)

If this is the case, it will be impossible to provides a real fix for this API (maybe change it). As the items might not be rendered inside the `FlatList`, we cannot compute the `deltaY` properly without specifying `getItemLayout` (But if it currently doesn't work when the layout size is given, it can be done!)

Don't forget that we could start at a specific item index with `initialScrollIndex` / have an `inverted ` list, meaning it might have N items before, which might never have been rendered. `autoScrollToTopThreshold` would be to compute without `getItemLayout` in such cases.

<img width="887" alt="Screenshot 2022-03-21 at 21 30 59" src="https://user-images.githubusercontent.com/1902323/159358687-b605ed7f-1a53-4c2b-a9e0-9afe8acb29a1.png">

**Comment by roryabraham** [comment]
@zoontek In regards to your last post and the question posed in the image: `How can we compute this since the items are currently not rendered and their height is variable`?

I could be wrong, but my understanding is that the dynamically computed height will be available in [`addUIBlock`](https://github.com/facebook/react-native/blob/8bd3edec88148d0ab1f225d2119435681fbbba33/React/Modules/RCTUIManager.h#L100), and since it's running in the UI thread we can synchronously adjust the `contentOffset` of the ScrollView before `addUIBlock` renders the new items.

**Comment by roryabraham** [comment]
I agree it's possible that the problem exists only with a VirtualizedList, but I'm still not sure why. My loose understanding of @azimgd's latest theory is that it has something to do with referential integrity. Something like this:

1. Let's say `VirtualizedList` has `data` with 1000 items is rendering items `[400, 600]`
1. `onStartReached` is called and item `400` is marked as the first visible item. Its position in the `ScrollView` is correctly recorded at `0px`.
1. Items `[300-399]` are added to the list. Right before we render them, we try to check the new position of item `400`.
1. Because `VirtualizedList` has added new items to the list, item `400` references a different JS object than it did before. Because the native views are recycled, the native layer can't tell the difference between item `400` and item `300`.

Honestly not sure, but that's my best attempt at understanding @azimgd's latest hypothesis for now.

**Comment by zoontek** [comment]
> I could be wrong, but my understanding is that the dynamically computed height will be available in [`addUIBlock`](https://github.com/facebook/react-native/blob/8bd3edec88148d0ab1f225d2119435681fbbba33/React/Modules/RCTUIManager.h#L100), and since it's running in the UI thread we can synchronously adjust the `contentOffset` of the ScrollView before `addUIBlock` renders the new items.

No, you are right, but it doesn't solve the main issue: by rendering items on scrolling to the top of the list, you shift  `autoscrollToTopThreshold` limit. This could be fixed by precising `getItemLayout` + fix the ScrollView module to support forced and consistant items layout size (currently, `VirtualizedList` is based on `ScrollView` native module (https://github.com/facebook/react-native/blob/0.68-stable/Libraries/Lists/VirtualizedList.js)

> 1. Let's say `VirtualizedList` has `data` with 1000 items is rendering items `[400, 600]`
> 2. `onStartReached` is called and item `400` is marked as the first visible item. Its position in the `ScrollView` is correctly recorded at `0px`.
> 3. Items `[300-399]` are added to the list. Right before we render them, we try to check the new position of item `400`.
> 4. Because `VirtualizedList` has added new items to the list, item `400` references a different JS object than it did before. Because the native views are recycled, the native layer can't tell the difference between item `400` and item `300`.

Let's say we set `maintainVisibleContentPosition: 400` and `autoscrollToTopThreshold: 50` in your exemple. By scrolling to the top, say we render items 3 per 3 and items are immediately unmounted once they exit the screen.

<img width="1256" alt="Screenshot 2022-03-21 at 23 39 30" src="https://user-images.githubusercontent.com/1902323/159374725-a1270f71-3bfd-4607-9dfa-ac44a22b2137.png">

As you see, `autoscrollToTopThreshold` will not be related to the "real" top of the list but will depend of how much items are rendered above, and their size. But if we are aware that the first rendered items in the list is and we know via `getItemLayout` items constant height, we could determine if we are in the autoscrollToTop threshold.

**Comment by azimgd** [comment]
`This is a draft`

## FlatList
<img src="https://user-images.githubusercontent.com/4882133/159477861-ffeeb15d-bdf2-4b94-bb0c-a8754959b0f6.jpg" width="360" />

```js
<ScrollView>
{head.map(renderItem)} → https://github.com/Expensify/react-native/blob/master/Libraries/Lists/VirtualizedList.js#L910-L926
{virtualized.map(renderItem)} → https://github.com/Expensify/react-native/blob/master/Libraries/Lists/VirtualizedList.js#L962-L969
{tail.map(renderItem)} → https://github.com/Expensify/react-native/blob/master/Libraries/Lists/VirtualizedList.js#L991-L998
</ScrollView>
```

Output of `NSLog(@"Subviews count: %lu", self->_contentView.subviews.count);` at `prependUIBlock`:

- render cycles of virtualizer items
- only 100 items are passed into FlatList (500-600)
- onStartReached is turned off
- onEndReached is turned off

```
2022-03-22 18:25:23.468276+0500 MVCPExample[66740:6211501] Subviews count: 13
2022-03-22 18:25:23.468414+0500 MVCPExample[66740:6211501] Subviews count: 13
2022-03-22 18:25:23.468551+0500 MVCPExample[66740:6211501] Subviews count: 13
2022-03-22 18:25:23.468674+0500 MVCPExample[66740:6211501] Subviews count: 13
2022-03-22 18:25:23.468751+0500 MVCPExample[66740:6211501] Subviews count: 13
2022-03-22 18:25:23.469204+0500 MVCPExample[66740:6211501] Subviews count: 13
2022-03-22 18:25:23.474187+0500 MVCPExample[66740:6211501] Subviews count: 13
2022-03-22 18:25:23.545705+0500 MVCPExample[66740:6211501] Subviews count: 13
2022-03-22 18:25:23.627980+0500 MVCPExample[66740:6211501] Subviews count: 23
2022-03-22 18:25:23.707817+0500 MVCPExample[66740:6211501] Subviews count: 33
2022-03-22 18:25:23.788797+0500 MVCPExample[66740:6211501] Subviews count: 43
2022-03-22 18:25:23.855413+0500 MVCPExample[66740:6211501] Subviews count: 53
2022-03-22 18:25:23.921377+0500 MVCPExample[66740:6211501] Subviews count: 63
2022-03-22 18:25:23.988802+0500 MVCPExample[66740:6211501] Subviews count: 73
2022-03-22 18:25:24.064840+0500 MVCPExample[66740:6211501] Subviews count: 83
2022-03-22 18:25:24.129288+0500 MVCPExample[66740:6211501] Subviews count: 93
2022-03-22 18:25:24.187206+0500 MVCPExample[66740:6211501] Subviews count: 95
```

## Fix One
https://github.com/Expensify/react-native/blob/master/React/Views/ScrollView/RCTScrollView.m#L922
```js
// add self->_scrollView.contentOffset.y > 0 

CGFloat y = self->_scrollView.contentOffset.y + bottomInset;
hasNewView = subview.frame.origin.y > y && y > 0;
```

This will cause https://github.com/Expensify/react-native/blob/master/React/Views/ScrollView/RCTScrollView.m#L925
```js
ii == self->_contentView.subviews.count - 1
```

To grab last subview until virtualizer has completed it's job with rendering actual items (around 4 component render cycles for the sample of 100 items)

## Fix Two

Save the content height in `prependUIBlock`
```js
[manager
      prependUIBlock:^(__unused RCTUIManager *uiManager, __unused NSDictionary<NSNumber *, UIView *> *viewRegistry) {
self->_prevContentSizeHeight = self->_scrollView.contentSize.height;
```

https://github.com/Expensify/react-native/blob/master/React/Views/ScrollView/RCTScrollView.m#L954
```js
CGFloat deltaY = self->_scrollView.contentSize.height - self->_prevContentSizeHeight;
```

Adjust content height when new items are prepended

## Proposal three
Pass `range` returned by `computeWindowedRenderLimits` at VirtualizedList into `ScrollView` to get frame of the actual first visible item.

**Comment by parasharrajat** [comment]
I am trying to understand your proposal @azimgd. Sorry for the delay.

**Comment by medzomethod** [comment]
I can get it fixed asap @mallenexpensify

**Comment by parasharrajat** [comment]
How will you fix it @medzomethod? Please read about the contributing guidelines from CONTRIBUTING.md.

**Comment by parasharrajat** [comment]
I tried to understand your proposal @azimgd but I didn't completely get the good sense of what are you trying to achieve by all of those changes? I know this is a draft but thought of sharing the review with you. 

I am new to native code but I can understand the logic. It would be helpful for me if you can also tell us what are those variables you referenced in your proposals. You reviewed the native code and understand it but they might not be known to others (like me). 

so e.g. `self->_contentView	` => renderedContent 
  `subview.frame` => `Item Dimensions`. 
  
Proposal one, two, etc will be better tagged as steps. 


My understanding is that you are trying to get the correct first visible item on the content view when new items are prepended. Please let me know if I am wrong. Thanks. I will try to learn and understand the changes as much as possible.

**Comment by azimgd** [comment]
There are multiple things need to be addressed in order to integrate `maintainVisibleContentPosition` with `FlatList`:
- prevent multiple `onStartReached` calls on initial render in this [PR](https://github.com/Expensify/react-native/pull/7) 
- fix `onCellLayout` calls inside `VirtualizedList.CellRenderer` component
- calculate correct coordinates for each FlatList item (aka `subview`)

## fix `onStartReached` calls
This prop is currently implemented under `maybeCallOnEdgeReached` [here](https://github.com/Expensify/react-native/blob/090e78d3f3b83b9a782290d7426dd0dba1e97af0/Libraries/Lists/VirtualizedList.js#L1520) which is called for following scenarios:
- `_onLayout`: good, check and execute when `ScrollView` is rendered. Allows us loading content in both directions on first mount.
- `_onScroll`: good, check and execute when `ScrollView` is scrolled.
- `_onContentSizeChange`: needs modification, since this method is tied to `_onLayout`.

We need to modify and prevent execution of `onStartReached` inside `ScrollView.onLayout` when offset is changed within native code because this happens for the single event:
1. calls `onStartReached` from `onLayout`
2. append content
3. calls `onStartReached` from `onContentSizeChange`
4. sets offset

### Possible solutions
1. pass callee name `maybeCallOnEdgeReached('onContentSizeChange')` and blacklist `onStartReached` when callee equals to _onContentSizeChange_
2. compare execution timestamps and don't call if the diff is just couple milliseconds (not the best)

## fix `onCellLayout` calls 
Need to actually investigate more why `fillRateHelper` is disabled on development, and `onCellLayout` from `VirtualizedList` on called inside `CellRenderer` resulting frame dimensions not being cached at `this._frames`.

Just changing [this](https://github.com/Expensify/react-native/blob/090e78d3f3b83b9a782290d7426dd0dba1e97af0/Libraries/Lists/VirtualizedList.js#L2105) to: ```const onLayout = this.props.onLayout;``` for now.

## calculate correct coordinates for each FlatList item (aka `subview`)
This is the most important we need to address. 

Coordinates for `minIndexForVisible` are calculated at `prependUIBlock` [here](https://github.com/facebook/react-native/blob/8bd3edec88148d0ab1f225d2119435681fbbba33/React/Views/ScrollView/RCTScrollView.m#L909)

This works in the following way, say we have a ScrollView with 5 children and want to get the position of children #3:
```js
<ScrollView>
  <View key={1}>
  <View key={2}>
  <View key={3}>
  <View key={4}>
  <View key={5}>
</ScrollView>
```

Should be easy enough, we can iterate through the children of ScrollView and pick 3rd item using this code:
```js
UIView *subview = nil
for (NSUInteger ii = 0; ii < self->_contentView.subviews.count; ++ii) {
  // dummy condition ofc, calculation could be changed according to offsets or whatever
  if (ii === 2) subview = self->_contentView.subviews.count[ii]
```

However this approach of iterating through `self->_contentView.subviews.count` will break for Flatlist due to:
- Virtualiztion
- Rendering supplement components like `ListHeaderComponent`, `ListFooterComponent`, `ListEmptyComponent`, `ItemSeparatorComponent`
- `_computeBlankness` and rendering long empty components that fill empty offset.

example:
```js
// virtualized state = { first: 3, last: 4 }
<ScrollView>
  <View key={1}> //always rendered because this is head
  // emptiness
  <View key={3}>
  <View style={{ height: 1000px }}> // blank view
  <View key={5}> //always rendered because this is tail
</ScrollView>
```

Having offsets for minIndexForVisible calculations on Obj-C side is not realistic without changing API for me. But, since we already have all frame sizes calculated at VirtualizedList, this could be simply passed into native code.

```js
// this.frames = { 0: {height: 50, start: 0, end: 50}, 1: {height: 150, start: 50, end: 200} }
<View key={0} style={{ height: 50 }}>
<View key={1} style={{ height: 150 }}>
```

Extend `maintainVisibleContentPosition` [here](https://github.com/Expensify/react-native/blob/090e78d3f3b83b9a782290d7426dd0dba1e97af0/Libraries/Lists/VirtualizedList.js#L1274) to include new `frames` object.

```js
<ScrollView
  {...props}
  maintainVisibleContentPosition={{
    ...props.maintainVisibleContentPosition,
    frames: {...this._frames},
  }}
```

then we can change our obj-c code:
```diff
- for (NSUInteger ii = minIdx; ii < self->_contentView.subviews.count; ++ii) {
+ for (NSUInteger ii = minIdx; ii < [self->_maintainVisibleContentPosition[@"frames"] count] - 1; ++ii) { 
```

This is the easiest solution I could come up that for now with the least amount of modifications required. Other solutions like simply setting offset to `contentSize - prevContentSize` definitely works and could be used if you are not really planning on using `minIndexForVisible` and simply having persistent scroll on appending items are described in my comment above https://github.com/Expensify/App/issues/7925#issuecomment-1075208622

@roryabraham @parasharrajat

**Comment by parasharrajat** [comment]
Awesome, I will check it tomorrow.

**Comment by parasharrajat** [comment]
Couldn't test it. Upgrading the macOS version on my Mac has broken the IOS simulator. I am trying to fix it.

**Comment by parasharrajat** [comment]
I am not familiar with the inner workings of RN FlatList and it seems quite complex to understand quickly. Also, I am not good at native code. I might take some time to really understand the suggestions here. Looking at the urgency of the issue, I would request someone else to take over.  

@roryabraham Could you please continue the discussion here?

**Comment by roryabraham** [comment]
Thanks for all your investigation @azimgd. I unfortunately have not had time to fully investigate your latest proposal yet, but I am hoping to do that in the next few days and get back to you. Thanks.

**Comment by azimgd** [comment]
Adjusting `state<{first, last}>` at `VirtualizedList` fixes `maintainVisibleContentPosition` prop on ios.

`state<{first, last}>` aka range, is holding the windowed render limits result.
```js
→ data.length = 1000
→ {first: 0, last: 150} // renders only 0-150, virtualizes 150-1000

→ data.length = 2000
→ {first: 1000, last: 1150} // renders only 1000-1150, virtualizes 0-1000, 1150-2000
```

once items are prepended into `<FlatList data={items} />` we should simply calculate amount of items prepended and shift range to make `maintainVisibleContentPosition` work.

you can observe it by hardcoding range at [VirtualizedList#getDerivedStateFromProps:](https://github.com/facebook/react-native/blob/main/Libraries/Lists/VirtualizedList.js#L773)
```js
static getDerivedStateFromProps(newProps: Props, prevState: State): State {
    return (() => {
      const perPage = 1000
      const itemCount = newProps.data.length
      if (itemCount === 1 * perPage) return { first: 0, last: 150 }
      if (itemCount === 2 * perPage) return { first: 1 * perPage - 100, last: 1 * perPage + 100 }
      if (itemCount === 3 * perPage) return { first: 2 * perPage - 100, last: 2 * perPage + 100 }
      if (itemCount === 4 * perPage) return { first: 3 * perPage - 100, last: 3 * perPage + 100 }
      if (itemCount === 5 * perPage) return { first: 4 * perPage - 100, last: 4 * perPage + 100 }
    })()
}
```

Proposal is to inject `customRangeExtractor` function into `getDerivedStateFromProps` and invoke it from there, e.g.:
```js
const customRangeExtractor = (dataLength, first, last) => {
  const itemsPerPage = 1000; // calculate items per page and prepended items here

  for (let i = itemsPerPage + itemsPerPage; i <= dataLength; i = i + itemsPerPage) {
    if (i !== dataLength) continue;
    if (i === diff) continue;
    const nextFirst = first + i - itemsPerPage
    const nextLast = last + i - itemsPerPage
    return { first: nextFirst, last: nextLast, diff: i }
  }


  return null
}

static getDerivedStateFromProps(newProps: Props, prevState: State): State {
  return customRangeExtractor()
}
```

**Comment by roryabraham** [comment]
@azimgd I think your latest proposal gets pretty close to the core of the issue – though I'm not sure about the `customRangeExtractor` solution. Would the developer implementing the bidirectional `FlatList` have to provide that `customRangerExtractor` implementation themselves?

I think we'll be able to implement a generic solution in VirtualizedList that solves the range problem ... I'm investigating to see what that might look like.

**Comment by azimgd** [comment]
@roryabraham Even though we could allow passing by props (might be helpful) I imagine that `defaultRangeExtractor` will be defined at `VirtualizedList` or `VirtualizedUtils`.

**Comment by roryabraham** [comment]
Hmmm, I guess that makes sense (though I'm not sure what use-cases there might be to pass in the `customRangeExtractor` via props). But it seems in line with the way other features have been implemented in the `VirtualizedList` component, so that sounds good to me.

I think we need to agree on the `defaultRangeExtractor` and how it will be implemented in `VirtualizedList` though, as that's a substantial piece of the solution.

**Comment by janicduplessis** [comment]
I've done a round of investigation on this issue and here are my findings. There are actually 2 separate issues:

### Issue 1

 `autoscrollToTopThreshold` calculation is wrong. This is pretty easy to fix and can be reproduced in the simple scrollview example. The problem is that the y value is set before adjusting content offset so it doesn't need to be adjusted with the delta. Here's the diff with the fix: https://github.com/facebook/react-native/compare/main...janicduplessis:patch-14.

### Issue 2

`maintainVisibleContentPosition` is broken when using virtualization and the new content pushes visible content outside its "window". This can be reproduced in this diff https://github.com/Expensify/react-native/pull/7/files#diff-e30adf9ac66888db4fe1a9c256590a37677724935458dae6b7a61b99331e292eR46-R48. When using a large PAGE_SIZE (which will cause a lot of items to be added) it will always push visible content outside of the list "window" which will cause currently visible views to be unmounted so the implementation of `maintainVisibleContentPosition` can't adjust the content inset since the visible views no longer exist.

The first illustration shows the working case, when the new content doesn't push visible content outside the window. The red box represents the window, all views outside the box are not mounted, which means the native implementation of `maintainVisibleContentPosition`  has no way to know it exists. In that case the first visible view is #2, after new content is added #2 is still inside the window so there's not problem adjusting content offset to maintain position. As you can see Step 1 and 3 result in the same position for all initial views.

The second illustation shows the broken case, when new content is added and pushes the first visible view outside the window. As you can see in step 2 the view #2 is no longer rendered so there's no way to maintain its position.

The third illustration shows the potential fix, we need to avoid adjusting the window when `maintainVisibleContentPosition`  this way the first visible view is not unmounted and content position can be maintained. I think this would be possible to implement by making `VirtualizedList` aware of `maintainVisibleContentPosition` and avoid updating its window when new content is added.

![image](https://user-images.githubusercontent.com/2677334/163263472-eaf7342a-9b94-4c49-9a34-17bf8ef4ffb9.png)

![image](https://user-images.githubusercontent.com/2677334/163263528-a8172341-137e-417e-a0c7-929d1e4e6791.png)

![image](https://user-images.githubusercontent.com/2677334/163263550-0b9aa391-f239-443c-9b7d-569e2ed1b550.png)

**Comment by azimgd** [comment]
> though I'm not sure what use-cases there might be to pass in the customRangeExtractor via props
Probably not too much indeed, as flatlist is quite customizable and allows passing custom props like overscan, windowsize etc.

I believe last function called `customRangeExtractor` defined at https://github.com/Expensify/App/issues/7925#issuecomment-1095459409 outlines my idea, if you would like to see a full solution before making a further decision I can work towards that as well.

**Comment by janicduplessis** [comment]
I think the fix can live in `VirtualizedList` directly as it will require knowing about a lot of internal state to apply the proper window adjustments. I don't see any uses for adding it as a prop.

**Comment by roryabraham** [comment]
> I believe last function called customRangeExtractor defined at https://github.com/Expensify/App/issues/7925#issuecomment-1095459409 outlines my idea, if you would like to see a full solution before making a further decision I can work towards that as well.

Can you please attempt to explain in plain English (without using code) how you will fix this problem directly and generically in `VirtualizedList`?

It's difficult to evaluate a proposal based off that code, because it's using hard-coded values. Also it just doesn't really make sense to me... the for loop doesn't do anything unless `i === itemCount`, so why not just reference that directly?

```diff
const customRangeExtractor = ({ itemCount, first, last }) => {    
-    const itemsPerPage = 1000
-    for (let i = itemsPerPage + itemsPerPage; i <= itemCount; i = i + itemsPerPage) {
-        if (i !== itemCount) continue;
-        const nextFirst = i - itemsPerPage
-        const nextLast = i - itemsPerPage + 150
-        return { first: nextFirst, last: nextLast }
-    }
-    return { first, last }
+    const itemsPerPage = 1000;
+    return itemCount % itemsPerPage === 0
+        ? {first: itemCount - itemsPerPage, last: itemCount - itemsPerPage + 150}
+        : {first, last};
}

static getDerivedStateFromProps(newProps: Props, prevState: State): State {
    return customRangeExtractor({ itemCount: newProps.data.length, ...prevState })
}
```

How will you determine `itemsPerPage`, and where did `150` come from? When would `itemCount % itemsPerPage` not be `0`?

**Comment by janicduplessis** [comment]
Here's my proposal for the issue described [here](https://github.com/Expensify/App/issues/7925#issuecomment-1098449646).

## Proposal

To fix `maintainVisibleContentPosition` when using `VirtualizedList` we need to make sure the visible items stay rendered when new items are added at the start of the list.

In order to do that we need to do the following:

- Detect new items that will cause content to be adjusted
- Adjust `{first,last}` state the the number of new items added at the start of the list
- Ignore certain updates while scroll metrics are invalid

### Detect new items that will cause content to be adjusted

We need to find how many items have been added before the first visible one. To do that we save the key of the first visible item in state `firstItemKey`. When data changes we can find the index of `firstItemKey` in the new data and that will be the amount we need to adjust the window state by.

In practice we could also simply save the first item in the list instead of the first visible item, as tracking visibility is more complex. For the case of bidirectional pagination this will work as items are only added at the start or end of the list.

### Adjust `{first,last}` state the the number of new items added at the start of the list

Once we have the adjusted number we can save this in a new state value `maintainVisibleContentPositionAdjustment`. Note that this needs to be in a separate state and cannot update `{first,last}` directly as other updates can happen between window adjustment and when react-native is done rendering.

This state is then cleared when we receive updated scroll metrics, once the native implementation is done adding the new items and adjusting the content offset.

This value is also only set when `maintainVisibleContentPosition` is set so this makes sure this maintains the currently behavior when that prop is not set.

### Ignore certain updates while scroll metrics are invalid

While the `maintainVisibleContentPositionAdjustment` state is set we know that the current scroll metrics are invalid since they will be updated in the native `ScrollView` implementation. In that case we want to prevent certain code from running.

One example is `onStartReached` that will be called incorrectly while we are waiting for updated scroll metrics.

## Implementation

I have a draft implementation that covers most of this proposal. It is on top of https://github.com/Expensify/react-native/pull/7.

https://github.com/Expensify/react-native/pull/8

It currently doesn't properly implement the first point and assumes all items are added at the start of the list. This causes windowing to be broken when scrolling down, until the next scroll event happens.

## Demo

https://user-images.githubusercontent.com/2677334/163293725-6e65aaf7-9e1b-462d-91c6-772054ad7b49.mov

**Comment by azimgd** [comment]
Just tested the changes on my local branch and this works pretty well. This is a modified version of the concept I mentioned above.

1. Calculate diff of prepended items by comparing `FlatList#keyExtractor`
2. Render items with new virtualized range with increased `first` and `last` values by diff
3. Let the `maintainVisibleContentPosition` shift the list and emit `onScroll` event
4. Reset the diff 

```js
static getDerivedStateFromProps(newProps: Props, prevState: State): State {

  ...

  // calculate diff of appended items
  let shiftMetrics = prevState.shiftMetrics || {}
  if (!shiftMetrics?.firstItem && getItemCount(data)) {
    shiftMetrics.firstItem = data[0]
  } else if (shiftMetrics?.firstItem && shiftMetrics?.firstItem !== data[0]) {
    // @TODO: add key extractor
    shiftMetrics.diff = data.findIndex(item => item === shiftMetrics?.firstItem)
  }
  
  // ensure component is not updated by shiftMetrics reset by returning null
  
  return {
    first: first + (shiftMetrics.diff || 0),
    last: last + (shiftMetrics.diff || 0),
  }
```

```js
_onScroll = (e: Object) => {
  // add an extra check to ensure no race-condition happen between actual scroll
  // and the side-effect of the content shift emitted by maintainVisibleContentPosition
  this.setState(state => ({
    ...state,
    shiftMetrics: {}
  }))
```

**Comment by roryabraham** [comment]
> we could also simply save the first item in the list instead of the first visible item, as tracking visibility is more complex. For the case of bidirectional pagination this will work as items are only added at the start or end of the list

What would be involved in tracking visibility? I have a couple questions that make me think using the first item in the list might not be ideal:

1. What if we have a large `onStartReachedThreshold`, such that the first item on the list is not visible when we call `onStartReached` and start loading new content? Would that cause a scroll jump up to what was previously the first item in the list?
1. What if we pass a non-zero `minIndexForVisible` parameter in the `maintainVisibleContentPosition` prop? Would we just use the Nth item in the list instead of the first one?

**Comment by roryabraham** [comment]
> For the case of bidirectional pagination this will work as items are only added at the start or end of the list.

Ideally we wouldn't rely on this assumption. I agree it's unlikely that content will be added in the middle of a list, but is it possible that deletion of content might simulate this case?

1. Client A has items 10-20 loaded.
1. Client B deletes items 15 and 17.
1. Client A reaches their start threshold, and fetches items 0-9. Items 15-17 are no longer valid.

Not sure if it's a problem, but it's something we should be aware of. We should explicitly call out any assumptions we make in documentation updates.

**Comment by roryabraham** [comment]
> keys need to be stable, and just using index won't work

Again, not sure if it will be a problem for us or not, but we should call this out in documentation updates. Also, can we succinctly define what is meant by "stable"?

In our case, the list items are called "report actions", and when they're created we first generate an optimistic `clientReportActionID` and `clientSequenceNumber`. When the write response comes back from the API, we overwrite the optimistic `reportActionID and `sequenceNumber` with the "canonical" one returned from our servers. In that regard, neither value is necessarily "stable".

**Comment by roryabraham** [comment]
> Once we have the adjusted number we can save this in a new state value maintainVisibleContentPositionAdjustment

How will this state value be used to extend the render window?

> This state is then cleared when we receive updated scroll metrics

How will clearing this state shrink the render window back to the expected range?

**Comment by roryabraham** [comment]
@azimgd I think your latest proposal is conceptually very similar to @janicduplessis's latest proposal. Both proposals:

1. Use item keys to determine the number of prepended items.
1. Save that adjustment in a separate state variable.
1. Adjust `first` and `last`, extending the range to include not only the prepended items, but also the first item before the new items were prepended.
1. After the scroll metrics update, reset the separate state used to adjust `first` and `last` in `onScroll`.

However, there are some notable differences:

1. Your proposal is missing an important piece that's called out in @janicduplessis's: skipping `maybeCallOnEdgeReached` if the adjustment state is non-null.
1. @janicduplessis did a much better job breaking down the problem and solution. To someone not well-versed with this issue, it might not be immediately clear that the proposals are similar. I had to unpack your solution by studying the code snippet, and without @janicduplessis's explanation above, I'm not sure I would've understood how it solves the problem.

**Comment by janicduplessis** [comment]
> What would be involved in tracking visibility? I have a couple questions that make me think using the first item in the list might not be ideal:

Visibility tracking should be relatively straightforward as VirtualizedList has context for scroll metrics and all item positions. However I'm not sure it is needed. Basically the goal of this piece of code is to determine if the native implementation of maintainVisibleContentPosition will be triggered, and by how many our items have been moved by new items added before our virtualization window (first state).

Detecting items added at the start of the list is good enough for bidirectional pagination, but of course will break if items are added at other positions.

1. I don't think this matters as VirtualizedList has knowledge of all items, even those not rendered. I can test to confirm, but I did play with that value a bit and seemed fine.

2. Right now we don't consider this value, I think its main use is to ignore changes to header views, which VirtualizedList separate from data, so as far as I know it is fine to ignore, at least when using first item to detect changes.

> Ideally we wouldn't rely on this assumption. I agree it's unlikely that content will be added in the middle of a list, but is it possible that deletion of content might simulate this case?

Good point, I haven't tested how `maintainVisibleContentPosition` behaves when removing items, we just need to make sure the JS implementation does the same thing so if content will be adjusted by the native implementation we also adjust the VirtualizedList window. In practice this probably won't break unless deleting a large amount of items at once that would cause a shift that moves visible content outside the virtualization window.

Right now when using the first item, a delete in the middle would cause no changes to the virtualization window since the first item did not change, which I think might be correct.

> Again, not sure if it will be a problem for us or not, but we should call this out in documentation updates. Also, can we succinctly define what is meant by "stable"?

This could be added to documentation for `maintainVisibleContentPosition` of VirtualizedList. By stable I mean keys are id based and not index based. This is already the preferred way to handle keys in VirtualizedList, but we could detect it and show a warning if used incorrectly.

So for example when adding new items key should be a unique id for the item like:

Good, id based key:
```js
// Good we can detect where new item was added.
[{key: 'a'}, {key: 'b'}] -> [{key: 'c'}, {key: 'a'}, {key: 'b'}]
```
Bad, index based key:
```js
// Looks like item was added at the end, even if it was added at the start since all key shifted.
[{key: 1}, {key: 2}] -> [{key: 1}, {key: 2}, {key: 3}]
```

**Comment by janicduplessis** [comment]
> In our case, the list items are called "report actions", and when they're created we first generate an optimistic clientReportActionID...

I think this would work still, it would be more like an item update / replace. In this case the code to find first item will not detect any change, or detect the first item no longer exist, which both case result in no change to windowing. Should add this in the list of things to test though!

> How will this state value be used to extend the render window?

It is added to the first / last state in render (https://github.com/Expensify/react-native/pull/8/files#diff-7481ec8ed5532798b075df637e805a8e439807aa2ce671208c24068c286361e8R977).

> How will clearing this state shrink the render window back to the expected range?

This window in not actually bigger, but shifted. Once we have received a scroll event we have the proper scroll position that has been updated by the native implementation of `maintainVisibleContentPosition`. This means that the next time we calculate the window bounds (first/last) it will result in the same as when we were shifting it using maintainVisibleContentPositionAdjustment.

Here's an example of how events happen:

```
state: {first: 5, last: 10}
js scroll position: 100
rendered window: {first: 5, last: 10}

onStartReached, new item added at start
detect 1 new item added, set maintainVisibleContentPositionAdjustment to 1

rendered window: {first: 6, last: 11} // (5 from state + 1 from maintainVisibleContentPositionAdjustment)
js scroll position: 100 // this is currently invalid since the native implementation has updated this value but we haven't received that update in JS yet
prevent updates since we have invalid js scroll position

scroll event, js scrollPosition updated to 110 (whatever adjustement was needed to maintain visible item after adding 1)
set maintainVisibleContentPositionAdjustment to null, recalculate window state based on new scroll position
state: {first: 6, last: 11}
rendered window: {first: 6, last: 11}
```

**Comment by roryabraham** [comment]
Okay @janicduplessis I'm going to assign this issue to you. The next steps will be:

1. Complete implementation of https://github.com/Expensify/react-native/pull/8. There are some changes that are required that you've mentioned already. It would also be good to add some automated tests if possible (to improve the likelihood of Meta merging the changes upstream).
1. Merge https://github.com/Expensify/react-native/pull/7 and https://github.com/Expensify/react-native/pull/8 into the main branch in our repo.
1. Update this repo to use the forked version of React Native, test bidirectional scrolling in-product (on iOS and Web, Android will be addressed in a follow-up because `maintainVisibleContentPosition` is not yet supported)
1. Open up a PR against the upstream React Native repo with all the changes from our fork, along with required changes to the documentation site.

**Comment by janicduplessis** [comment]
Updated https://github.com/Expensify/react-native/pull/8 to use the first item key as indicator of items being added, the bidirectional paging example now works properly when scrolling in both directions.

**Comment by janicduplessis** [comment]
Hey @mallenexpensify!

I'm currently working with @roryabraham on https://github.com/Expensify/react-native/pull/8. Currently addressing minor review comments.

**Comment by roryabraham** [comment]
PR was just merged into the bidirectional-scrolling feature branch!

**Comment by roryabraham** [comment]
Putting this on hold while we work. It seems like this particular bug is fixed, but we'll keep this open until the fix is verified in our app.

**Comment by mallenexpensify** [comment]
https://github.com/Expensify/App/issues/7860 is being held on this.

**Comment by mallenexpensify** [comment]
Rory's been out all week, should have an update next week

**Comment by roryabraham** [comment]
Made some great progress on Android implementation last week, posted update [here](https://expensify.slack.com/archives/C035J5C9FAP/p1655407541640779)

**Comment by mallenexpensify** [comment]
Asking for an update I can post in here for visibility

**Comment by roryabraham** [comment]
Still working on some pieces here, making slow progress

**Comment by roryabraham** [comment]
Haven't really made progress the last two weeks. Hoping to push this forward in the second half of this week.

**Comment by roryabraham** [comment]
Lots of recent updates in the expensify-margelo slack room

**Comment by roryabraham** [comment]
Fork was published yesterday, we are trying to upgrade E/App to use it in https://github.com/Expensify/App/pull/9841

**Comment by roryabraham** [comment]
This fix has landed along with all the other bidirectional pagination features in main. All that remains is to see them used in action via a comment linking implementation. Moving this to monthly because it might be a few weeks before that's done.

**Comment by mallenexpensify** [comment]
Thanks for the update @roryabraham , couple things... 
1. [Pagination seems broken for chats #7860](https://github.com/Expensify/App/issues/7860) has been on hold pending this issue, should it still be on hold? 

> This fix has landed along with all the other bidirectional pagination features in main. All that remains is to see them used in action via a comment linking implementation. 

When should payment be issued?  If the fix can be tested, I wouldn't think we'd need to wait "to see them used in action via a comment linking implementation" in order to pay for the work being done (that said... my level of expertise here is low)

**Comment by roryabraham** [comment]
@mallenexpensify as far as I know, we can't easily test the bidirectional pagination features in Expensify/App without building comment linking or some other test/feature using the onStartReached prop.

That said, we'll be meeting with Margelo next week to discuss next steps. https://expensify.slack.com/archives/C035J5C9FAP/p1661412756144339

**Comment by friedolinfoerder** [comment]
Hey everyone, I am a developer at [Steuerbot](https://www.steuerbot.com/) and me and our frontend team are struggling with this issue for a long time, because our app needs to hold the position even if new elements were added at the top of the chat list.
For that reason we hoped an official android support for the `maintainVisibleContentPosition` property would be developed, but this never happened 😕 

So we tried many different other approaches: Clientside-only hacks, patching RN-Core and other things. Finally we came up with a solution, which fixes the issue on our side 🚀 
What did we do? We are providing our own version of the Android RN ScrollView implementation (overriding functions from the RN class). This way we can correct the scroll position based on information the JS side provides us (this works also during a "fling animation" by restarting the fling with corrected position!).

I sharing the link to our developed RN module with you; maybe this will work for your project too 😃 
https://github.com/steuerbot/react-native-bidirectional-flatlist

Just give it a try. You can use it as a FlatList-Replacement. We are looking forward to your feedback!

**Comment by roryabraham** [comment]
Thanks! We've actually completed a bidirectional pagination solution ourselves in our React Native fork, available [on npm](https://www.npmjs.com/package/@expensify/react-native). @janicduplessis [implemented `maintainVisibleContentPosition` directly in our React Native fork](https://github.com/Expensify/react-native/pull/14/files), and we're hoping to open PRs to upstream these features soon.

**Comment by JmillsExpensify** [comment]
We're kicking off comment linking, so hopefully we'll have an update on this issue in the coming weeks.

**Comment by roryabraham** [comment]
Still on HOLD for comment linking, which is WIP

**Comment by mallenexpensify** [comment]
@roryabraham , I'm OOO for a month, no assigning a BZ cuz our process for this issue is an edge case.  Unassign me and add the `Bug` label if you need someone to take any BZ actions

**Comment by JmillsExpensify** [comment]
Btw, comment linking is on the horizon at this point. Internal design is happening.

**Comment by mallenexpensify** [comment]
Is there a link to comment linking discussions or a doc being worked on?

**Comment by roryabraham** [comment]
- issue: https://github.com/Expensify/Expensify/issues/147480
- doc: https://docs.google.com/document/d/1v-ZaIRTZL5LIsyPWB0IopBNNgCajf5WC1OA8cVKtd8I/edit#heading=h.ns12dr1f35rx

**Comment by JmillsExpensify** [comment]
Not sure if C+ is explicitly required at this point, I mainly wanted to clear up that we have a plan that we are working on internally, thus I update the issue with the `internal` label appropriately.

**Comment by roryabraham** [comment]
Upstream PR is in review: https://github.com/facebook/react-native/pull/35993

**Comment by mallenexpensify** [comment]
@janicduplessis can you reply to the comments in the PR - https://github.com/facebook/react-native/pull/35993#issuecomment-1455683366

Pasting here in case others have ideas 

> I was able to take a look through the code, and description. I have some implementation comments, but before I leave those, I wanted to check to see if I understand the overall problem, and the viability of an alternative method of doing this.
> 
> From the flow, here, it seems like:
> 
> User adds data before the viewport
> VirtualizedList responds to the new data, by rendering an index-based window around the new data
> maintainVisibleContentPosition doesn't save us and adjust scroll position because the old item isn't still rendered
> What I am wondering, could we attack https://github.com/facebook/react-native/issues/2 by changing how VirtualizedList handles data changes. I.e. fix existing logic in getDerivedStateFromProps() to align by key instead of index when determining the next render range on prop change.
> 
> That would avoid needing to keep an arbitrarily large range of cells rendered, because the VirtualizedList decided to move its own point of reference in response to items shifting index.

and

> A comment pending on this implementation, that I think would effect that one as well, is that we usually want to avoid the full scan of data like this change is currently doing. The model is meant to allow data to be backed by something expensive/lazy, so doing a full search throughout all data defeats that.
> 
> Though I am not sure the current API provides a way to diff before/after efficiently, without searching for keys. But we could also maybe optimize that search (e.g. look around the last index first).

**Comment by mallenexpensify** [comment]
From @janicduplessis [last week](https://github.com/facebook/react-native/pull/35993#issuecomment-1493364229)
>  I will also be away for the next few weeks so I will work on addressing your last round of feedback when I am back.

**Comment by janicduplessis** [comment]
All of the PRs have now landed upstream!

**Comment by mallenexpensify** [comment]
It looks like https://github.com/facebook/react-native/pull/35993#issuecomment-1493364229 is closed, is there anything else we're waiting on here?

**Comment by roryabraham** [comment]
It's just a documentation update, but I think the last PR we want to see merged before this is 100% complete is https://github.com/facebook/react-native-website/pull/3721

**Comment by darkbasic** [comment]
Well I guess that was quick :)

**Comment by roryabraham** [comment]
https://github.com/facebook/react-native-website/pull/3721 merged. Closing this out!

**Comment by mallenexpensify** [comment]
I'm assuming no contributors are due compensation via Upwork.  Please comment if  you disagree.
